### PR TITLE
[MKT-305]: feat/integrated checkout

### PR DESCRIPTION
**Pull request by xabg2**
Integrated checkout in drive web with registration + product payment on the same page. 
For now, it only supports payments for B2C. We continue to use Stripe Checkout for B2B plans.

### CODE REVIEW

**Comment by xabg2** [comment]
⚠️ To do before merging the PR:

- [x] Merge main branch
- [x] Add address billing inputs to the checkout
- [x] Merge the SDK PR (The PR is not open yet)
- [x] Merge the payments PR (merge from main and modify things if necessary)

File: src/app/payment/types.ts [hunk 1]
```diff
@@ -60,4 +68,85 @@ export enum LifetimeTier {
...
+export const THEME_STYLES = {
+  dark: {
+    backgroundColor: 'rgb(17 17 17)',
+    textColor: 'rgb(255 255 255)',
+  },
+  light: {
+    backgroundColor: 'rgb(255 255 255)',
+    textColor: 'rgb(17 17 17)',
+  },
+};
```
Comment (CandelR) [inline src/app/payment/types.ts]: This not seems a payment type,  I would move it to UI related file.

File: src/app/payment/types.ts [hunk 2]
```diff
@@ -60,4 +68,85 @@ export enum LifetimeTier {
...
+    applePay: 'auto',
+    googlePay: 'auto',
+  },
+  layout: {
+    type: 'accordion',
+    defaultCollapsed: false,
+    radios: false,
+    spacedAccordionItems: true,
+  },
+};
```
Comment (CandelR) [inline src/app/payment/types.ts]: This seems a configuration values, move it from types file.

File: src/App.tsx [hunk 3]
```diff
@@ -188,7 +188,7 @@ const App = (props: AppProps): JSX.Element => {
               to={`/?preferences=open&section=account&subsection=${params.get('tab') ?? 'account'}`}
             />
             <Redirect from="/app/:section?" to={{ pathname: '/:section?', search: `${queryParameters}` }} />
-            {pathName !== 'checkout-plan' && isMobile && isAuthenticated ? (
+            {pathName !== 'checkout-plan' && pathName !== 'checkout' && isMobile && isAuthenticated ? (
```
Comment (CandelR) [inline src/App.tsx:191]: What is the difference between the paths `checkout-plan` and `checkout`? Must they coexist?

File: src/app/drive/types/index.ts [hunk 4]
```diff
@@ -13,6 +13,7 @@ export interface DriveFolderData {
   encrypt_version: string | null;
   icon: string | null;
   iconId: number | null;
+  parentUuid: string;
```
Comment (CandelR) [inline src/app/drive/types/index.ts]: Already exists `parentUuid` field in `DriveFolderData` interface

File: src/app/i18n/locales/en.json [hunk 5]
```diff
@@ -172,6 +172,57 @@
...
+      "total": "Total: ",
+      "addCoupon": {
+        "buttonTitle": "Add coupon",
+        "inputText": "Coupon code",
+        "applyCodeButtonTitle": "Apply"
+      },
+      "amountSaved": "Save ",
+      "withAnnualBilling": "with annual billing"
+    }
+  },
```
Comment (CandelR) [inline src/app/i18n/locales/en.json]: Remember to add the other translations 🇨🇳 🇫🇷 🇮🇹 🇩🇪 🇪🇸

File: src/app/i18n/provider/TranslationProvider.tsx [hunk 6]
```diff
@@ -14,7 +14,8 @@ interface TranslationProviderProps {
 export const TranslationProvider: React.FC<TranslationProviderProps> = ({ children }) => {
   const { t } = useTranslation();
 
-  const translateList = (key: string) => t(key, { returnObjects: true }) as string[];
+  const translateList = (key: string, props?: Record<string, unknown>) =>
+    t(key, { returnObjects: true, replace: props }) as string[];
```
Comment (CandelR) [inline src/app/i18n/provider/TranslationProvider.tsx:18]: what is the `replace` functionality?

File: src/app/payment/components/checkout/Header.tsx [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { ReactComponent as InternxtLogo } from 'assets/icons/big-logo.svg';
+
+export const HeaderComponent = () => (
+  <div className="flex w-full flex-row justify-between">
+    <div className="flex flex-row space-x-2">
+      <InternxtLogo className="h-auto w-28 text-gray-100" />
+      <p className="text-lg text-gray-70">Checkout</p>
```
Comment (CandelR) [inline src/app/payment/components/checkout/Header.tsx]: Doesn't `Checkout` have to have a translation?

File: src/app/payment/components/checkout/InputsComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import PasswordInput from 'app/auth/components/PasswordInput/PasswordInput';
+import TextInput from 'app/auth/components/TextInput/TextInput';
+import { IFormValues } from 'app/core/types';
+import { useTranslationContext } from 'app/i18n/provider/TranslationProvider';
+import { MAX_PASSWORD_LENGTH } from 'app/shared/components/ValidPassword';
+import { FieldErrors, UseFormRegister } from 'react-hook-form';
+
+interface InputsComponent {
```
Comment (CandelR) [inline src/app/payment/components/checkout/InputsComponent.tsx]: nit: `InputsComponent` -> `InputsComponentProps` to mantain consistency with the rest of the project code

File: src/app/payment/components/checkout/ProductCardComponent.tsx [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                <SealPercent weight="fill" size={24} />
+                <p className="">
+                  {translate('checkout.productCard.saving', {
+                    percent: couponCodeData?.percentOff,
+                  })}
+                </p>
+              </div>
+              <p className="text-gray-50 line-through">{normalPriceAmount}€</p>
+            </div>
+          ) : undefined}
```
Comment (CandelR) [inline src/app/payment/components/checkout/ProductCardComponent.tsx]: could use 
```suggestion
          {couponCodeData && (
            <div className="flex flex-row items-center justify-between font-semibold">
              <div className="flex flex-row items-center space-x-2 text-green-dark">
                <SealPercent weight="fill" size={24} />
                <p className="">
                  {translate('checkout.productCard.saving', {
                    percent: couponCodeData?.percentOff,
                  })}
                </p>
              </div>
              <p className="text-gray-50 line-through">{normalPriceAmount}€</p>
            </div>
          )}
```

File: src/app/payment/components/checkout/ProductCardComponent.tsx [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+          <p className="text-2xl font-bold text-gray-100">
+            {translate(`checkout.productCard.plan.${selectedPlan.interval}`, {
+              spaceToUpgrade: bytes,
+              interval: selectedPlan.interval,
+            })}
+          </p>
+          <div className="flex flex-row items-center justify-between text-gray-100">
+            <p className="font-medium">{translate(`checkout.productCard.billed.${selectedPlan.interval}`)}</p>
+            {/* TODO: Change currency if needed */}
+            <p className="font-semibold">{planAmount}€</p>
```
Comment (CandelR) [inline src/app/payment/components/checkout/ProductCardComponent.tsx]: Isn't the currency dynamic?

File: src/app/payment/components/checkout/UserAuthComponent.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+  const isUserSignedIn = authMethod === 'userIsSignedIn';
+  const isNotUserSignedIn = authMethod !== 'userIsSignedIn';
+
+  const AUTH_LABEL: Partial<Record<AuthMethodTypes, string>> = {
+    signIn: translate('auth.login.dontHaveAccount'),
+    signUp: translate('auth.signup.haveAccount'),
+  };
+
+  const CHANGE_AUTH_METHOD: Partial<Record<AuthMethodTypes, string>> = {
```
Comment (CandelR) [inline src/app/payment/components/checkout/UserAuthComponent.tsx]: Maybe with the tag _TEXT it is better to understand that they are texts?

File: src/app/payment/services/auth-checkout.service.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import { doLogin, getNewToken } from 'app/auth/services/auth.service';
+import localStorageService from 'app/core/services/local-storage.service';
+import { decryptPrivateKey } from 'app/crypto/services/keys.service';
+import { AppDispatch } from 'app/store';
+import { planThunks } from 'app/store/slices/plan';
+import { productsThunks } from 'app/store/slices/products';
+import { referralsThunks } from 'app/store/slices/referrals';
+import { initializeUserThunk, userActions, userThunks } from 'app/store/slices/user';
+
+const signUp = async (
```
Comment (CandelR) [inline src/app/payment/services/auth-checkout.service.ts:13]: Why have you generated other signup and login functions instead of using the existing ones?

File: src/app/payment/services/checkout.service.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+  if (response.status !== 200) {
+    throw new Error('Plan not found');
+  }
+
+  const data = await response.json();
+
+  return data;
+};
+
+const fetchPromotionCodeByName = async (promotionCode: string): Promise<CouponCodeData> => {
```
Comment (CandelR) [inline src/app/payment/services/checkout.service.ts]: this arg `promotionCode: string` is the promotion name or the promotion code?

File: src/app/payment/services/checkout.service.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+const getClientSecretForPaymentIntent = async (
+  customerId: string,
+  amount: number,
+  planId: string,
+  promoCode?: string,
+): Promise<ClientSecretData> => {
+  const { clientSecret: client_secret } = await paymentService.createPaymentIntent(
+    customerId,
+    amount,
+    planId,
```
Comment (CandelR) [inline src/app/payment/services/checkout.service.ts:52]: With `planId` is it possible to buy business or individual plans?

File: src/app/payment/services/payment.service.ts [hunk 15]
```diff
@@ -51,6 +52,31 @@ const paymentService = {
     return stripe;
   },
 
+  // async createCustomer(name: string, email: string): Promise<{ customerId: string }> {
+  //   const paymentsClient = await SdkFactory.getInstance().createPaymentsClient();
+  //   return paymentsClient.createCustomer(name, email);
+  // },
```
Comment (CandelR) [inline src/app/payment/services/payment.service.ts]: forgot to remove this commented code

File: src/app/payment/types.ts [hunk 16]
```diff
@@ -60,4 +68,85 @@ export enum LifetimeTier {
...
+
+export type Action =
+  | { type: 'SET_PLAN'; payload: PlanData }
+  | { type: 'SET_CURRENT_PLAN_SELECTED'; payload: CurrentPlanSelected }
+  | { type: 'SET_PROMO_CODE_NAME'; payload: string }
+  | { type: 'SET_COUPON_CODE_DATA'; payload: CouponCodeData }
+  | { type: 'SET_STRIPE'; payload: Stripe | null }
+  | { type: 'SET_ELEMENTS_OPTIONS'; payload: StripeElementsOptions }
+  | { type: 'SET_AUTH_METHOD'; payload: AuthMethodTypes }
+  | { type: 'SET_ERROR'; payload: PartialErrorState };
```
Comment (CandelR) [inline src/app/payment/types.ts]: I wouldn't put the related redux store types here, I would add it together with the related redux store files.

File: src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+  authMethod: AuthMethodTypes;
+  error?: PartialErrorState;
+  couponCodeData?: CouponCodeData;
+  upsellManager: {
+    isUpsellSwitchActivated: boolean;
+    showUpsellSwitch: boolean;
+    onUpsellSwitchButtonClicked: () => void;
+    amountSaved: number | undefined;
+    amount: number | undefined;
+  };
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx]: If this type is repeated, it is still convenient to extract it to use it in the different places where it is needed, so for possible future changes it will be easier and we avoid the possibility of modifying only one of the implementations

File: src/app/payment/views/IntegratedCheckoutView/checkoutReducer.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
+import { Action, State } from '../../../payment/types';
+
+export const initialStateForCheckout: State = {
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/checkoutReducer.ts]: I would move this file to its own directory, so you have everything related to redux in one place, types, store, actions, etc.

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+        type: 'SET_ELEMENTS_OPTIONS',
+        payload: {
+          ...(elementsOptions as any),
+          amount: plan![planType].amount,
+        },
+      });
+    },
+    isUpsellSwitchActivated,
+    showUpsellSwitch: !!plan?.upsellPlan,
+    amountSaved: plan?.upsellPlan ? (plan?.selectedPlan.amount * 12 - plan?.upsellPlan.amount) / 100 : undefined,
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: Avoid magic numbers, add it to a descriptive constant instead

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+
+  const upsellManager = {
+    onUpsellSwitchButtonClicked: () => {
+      setIsUpsellSwitchActivated(!isUpsellSwitchActivated);
+      const planType = isUpsellSwitchActivated ? 'selectedPlan' : 'upsellPlan';
+      dispatchReducer({ type: 'SET_CURRENT_PLAN_SELECTED', payload: plan![planType] });
+      dispatchReducer({
+        type: 'SET_ELEMENTS_OPTIONS',
+        payload: {
+          ...(elementsOptions as any),
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: `as any` not good, try to not use `any`

File: src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+                  onAuthMethodToggled={onAuthMethodToggled}
+                  userData={userData}
+                  onLogOut={onLogOut}
+                />
+                <div className="flex flex-col space-y-8 pb-20">
+                  <p className="text-2xl font-semibold text-gray-100">2. {translate('checkout.paymentTitle')}</p>
+                  <PaymentElement options={PAYMENT_ELEMENT_OPTIONS} />
+                  {error?.stripe && <div className="text-red-dark">{error.stripe}</div>}
+                  <Button type="submit" id="submit" className="hidden lg:flex">
+                    {isExecutingPaymentAndAuth && isValid ? 'Paying...' : translate('checkout.pay')}
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx]: forgot to translate `Paying...`

File: src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+              <div className="flex w-full max-w-lg flex-col space-y-5">
+                <ProductFeaturesComponent
+                  selectedPlan={selectedPlan}
+                  couponCodeData={couponCodeData}
+                  couponError={error?.coupon}
+                  onCouponInputChange={onCouponInputChange}
+                  upsellManager={upsellManager}
+                />
+                <Button type="submit" id="submit" className="flex lg:hidden">
+                  {isExecutingPaymentAndAuth && isValid ? 'Paying...' : translate('checkout.pay')}
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx]: Same here

File: src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+        clientSecret = client_secret;
+      }
+
+      const confirmIntent = type === 'setup' ? stripe.confirmSetup : stripe.confirmPayment;
+
+      const { error } = await confirmIntent({
+        elements,
+        clientSecret,
+        confirmParams: {
+          return_url: 'http://localhost:3000/checkout/success',
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx]: I think you forgot to change this, be careful with hardcoded urls.

File: src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+        },
+      });
+
+      if (error) {
+        handleError('stripe', 'Something went wrong while confirming payment. Try again.');
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx]: There are several untranslated texts, will multiple languages be supported or only English here?

File: src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+        });
+    }
+  }, [user]);
+
+  function onAuthMethodToggled(authMethod: AuthMethodTypes) {
+    handleAuthMethod(authMethod);
+    reset(undefined, { keepErrors: false, keepValues: false });
+  }
+
+  const handleCheckout: SubmitHandler<IFormValues> = async (formData, event) => {
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutView.tsx]: A complex function, try to break it down into smaller functions to avoid it becoming hell to maintain.
I also think that by having a wrapper you can avoid having this kind of functions in the view itself, and pass them through props to decouple the view from any functionality that is not related to managing the UI.

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+  const { authMethod, error, currentPlanSelected, plan, stripe, couponCodeData, elementsOptions, promoCodeName } =
+    state;
+  const { backgroundColor, textColor } = THEME_STYLES[checkoutTheme as string];
+
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search);
+    const planId = params.get('planId');
+    const promotionCode = params.get('promotion_code');
+
+    const elementsOptionsParams: StripeElementsOptions = {
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: I think there is no need for `elementsOptionsParams` to be inside the `useEffect`

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+  const { doRegister } = useSignUp('activate');
+
+  const [state, dispatchReducer] = useReducer(checkoutReducer, initialStateForCheckout);
+  const [isUpsellSwitchActivated, setIsUpsellSwitchActivated] = useState<boolean>(false);
+
+  const { authMethod, error, currentPlanSelected, plan, stripe, couponCodeData, elementsOptions, promoCodeName } =
+    state;
+  const { backgroundColor, textColor } = THEME_STYLES[checkoutTheme as string];
+
+  useEffect(() => {
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx:169]: This useEffect is too complex. 
It handles multiple responsibilities, you can choose to separate it into different smaller useEffect and have each one focus on handling one topic.
For example one for stripe configuration, one for getting the plan and one for getting the promo code.
You can also create auxiliary functions to extract parts of the code into functions to perform the necessary actions and have each one focus on a task.

**Review by CandelR — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 3 above)
Comment (xabg2) [reply src/App.tsx:191]: For now, yes,
/checkout-plan takes the user to the Stripe checkout and /checkout is the pathname for the integrated checkout. For now you should keep both because the integrated checkout is only available for b2c payments and not b2b.

File: src/app/drive/types/index.ts (hunk 4 above)
Comment (xabg2) [reply src/app/drive/types/index.ts]: Removed 👍🏽

File: src/app/i18n/locales/en.json (hunk 5 above)
Comment (xabg2) [reply src/app/i18n/locales/en.json]: I keep it in mind, adding the translations rn 👍🏽

File: src/app/i18n/provider/TranslationProvider.tsx (hunk 6 above)
Comment (xabg2) [reply src/app/i18n/provider/TranslationProvider.tsx:18]: It is used to replace the elements that are put in square brackets. Since we are modifying the t object, we have to specify which options we want to use (the t object does it by itself), but in this case we must specify it.

File: src/app/payment/components/checkout/Header.tsx (hunk 7 above)
Comment (xabg2) [reply src/app/payment/components/checkout/Header.tsx]: Yes, I forgot about it.

File: src/app/payment/services/auth-checkout.service.ts (hunk 12 above)
Comment (xabg2) [reply src/app/payment/services/auth-checkout.service.ts:13]: Because I only need the signup and the dispatches. But there are redirects and extra logic we don't need for the integrated checkout. 

See it in [SignUp](https://github.com/internxt/drive-web/blob/master/src/app/auth/components/SignUp/SignUp.tsx#L138) and [Login](https://github.com/internxt/drive-web/blob/master/src/app/auth/components/LogIn/LogIn.tsx#L129).

File: src/app/payment/services/checkout.service.ts (hunk 14 above)
Comment (xabg2) [reply src/app/payment/services/checkout.service.ts:52]: I have to take a look at the new way of acquiring plans to be sure as they have changed quite a bit payments-server since this was done.

File: src/app/payment/services/checkout.service.ts (hunk 13 above)
Comment (xabg2) [reply src/app/payment/services/checkout.service.ts]: It is the Promotion name. I've updated the name of the param.

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 28]
```diff
@@ -0,0 +1,425 @@
...
+          ...(elementsOptionsParams as any),
+          mode: plan?.selectedPlan.interval === 'lifetime' ? 'payment' : 'subscription',
+          amount: plan?.selectedPlan.amount,
+          currency: plan?.selectedPlan.currency,
+          payment_method_types: ['card', 'paypal'],
+        },
+      });
+
+      const stripe = await stripePromise;
+      dispatchReducer({ type: 'SET_STRIPE', payload: stripe });
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: It seems that it is not necessary to store Stripe in the reducer

**Review by CandelR — APPROVED** [review]

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 29]
```diff
@@ -182,13 +183,28 @@ const CheckoutViewWrapper = () => {
           return;
         }
         if (checkoutTheme && plan) {
+          if (promotionCode) {
+            handleFetchPromotionCode(plan.selectedPlan.id as string, promotionCode).catch((err) => {
+              const error = err as Error;
+              const errorMessage = error.message.includes('Promotion code with an id')
+                ? error.message
+                : 'Something went wrong, try again later';
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: Is there no other way to check this than by checking the erorr message? With a status code or something like that?

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 30]
```diff
@@ -204,11 +220,29 @@ const CheckoutViewWrapper = () => {
...
-      handleFetchPromotionCode(promoCodeName);
+    if (promoCodeName && currentSelectedPlan) {
+      handleFetchPromotionCode(currentSelectedPlan?.id, promoCodeName).catch((err) => {
+        const error = err as Error;
+        const errorMessage = error.message.includes('Promotion code with an id')
+          ? error.message
+          : 'Something went wrong, try again later';
+        setError('coupon', errorMessage);
+        errorService.reportError(error);
+        setPromoCodeData(undefined);
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: this logic is duplicated, try to extract it to make it easier to maintain

File: src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx [hunk 31]
```diff
@@ -204,11 +220,29 @@ const CheckoutViewWrapper = () => {
...
+      });
     }
   }, [promoCodeName]);
 
+  useEffect(() => {
+    if (state.error?.coupon || state.error?.auth || state.error?.stripe) {
```
Comment (CandelR) [inline src/app/payment/views/IntegratedCheckoutView/CheckoutViewWrapper.tsx]: Add the conditionail in a variable with a descriptive name in order to understand better the reason

**Review by CandelR — CHANGES_REQUESTED** [review]

**Review by CandelR — APPROVED** [review]
